tutor/hooks/actions.py

```diff
@@ -4,58 +4,83 @@
 import sys
 import typing as t
 
+from typing_extensions import ParamSpec
+
 from .contexts import Contextualized
 
-# Similarly to CallableFilter, it should be possible to refine the definition of
-# CallableAction in the future.
-CallableAction = t.Callable[..., None]
+P = ParamSpec("P")
+# Similarly to CallableFilter, it should be possible to create a CallableAction alias in
+# the future.
+# CallableAction = t.Callable[P, None]
 
 DEFAULT_PRIORITY = 10
 
 
-class ActionCallback(Contextualized):
+class ActionCallback(Contextualized, t.Generic[P]):
     def __init__(
         self,
-        func: CallableAction,
+        func: t.Callable[P, None],
         priority: t.Optional[int] = None,
     ):
         super().__init__()
         self.func = func
         self.priority = priority or DEFAULT_PRIORITY
 
     def do(
-        self, *args: t.Any, context: t.Optional[str] = None, **kwargs: t.Any
+        self,
+        *args: P.args,
+        **kwargs: P.kwargs,
     ) -> None:
-        if self.is_in_context(context):
-            self.func(*args, **kwargs)
+        self.func(*args, **kwargs)
 
 
-class Action:
+class Action(t.Generic[P]):
     """
-    Each action is associated to a name and a list of callbacks, sorted by
-    priority.
+    Action hooks are callbacks that are triggered independently from one another.
```

Comment:
An action has callbacks, but the action itself isn't a callback.
```suggestion
    Action hooks are callbacks that have triggered independently from one another.
```
Otherwise, great improvement on the docstring.